Modules/gcmodule.c

```diff
@@ -1940,6 +1940,88 @@ gc_get_freeze_count_impl(PyObject *module)
 }
 
 
+/*[clinic input]
+gc.is_immortal -> bool
+
+    instance: object
+        The instance to perform the immortal check on.
+
+Check if an object has been immortalized.
+[clinic start generated code]*/
+
+static int
+gc_is_immortal_impl(PyObject *module, PyObject *instance)
+/*[clinic end generated code: output=743a163cb6aaf384 input=815182ca5c5d81e4]*/
+{
+#ifdef Py_IMMORTAL_INSTANCES
+    return _Py_IsImmortal(instance);
+#else
+    return 0;
+#endif  /* Py_IMMORTAL_INSTANCES */
+}
+
+
+#ifdef Py_IMMORTAL_INSTANCES
+static int
+immortalize_object(PyObject *obj, PyObject *Py_UNUSED(ignored))
+{
+    _Py_SetImmortal(obj);
+    /* Special case for PyCodeObjects since they don't have a tp_traverse */
```

Comment:
This is okay I think but should we consider adding tp_traverse to PyCodeObject?

Reply:
That was my original approach but I walked back on it. Even though PyCodeObjects is a container, it will never have a circular reference. Thus, adding `tp_traverse` will increase the cost of a GC by now traversing all code objects and it will never collect. 

Instead, I opted have that cost in the immortalization step since it shouldn't be frequently happening! So both should work, but this should be "cheaper" in the general case

Reply:
> That was my original approach but I walked back on it. Even though PyCodeObjects is a container, it will never have a circular reference. Thus, adding `tp_traverse` will increase the cost of a GC by now traversing all code objects and it will never collect

What about implement `tp_traverse` but don't activate support for the GC by always returning 0 from `tp_is_gc`?

Reply:
> Even though PyCodeObjects is a container, it will never have a circular reference. Thus, adding `tp_traverse` will increase the cost of a GC by now traversing all code objects and it will never collect.

That's probably true for code objects created by Python itself.  However, you can create cycles by calling types.CodeType() and passing the right kind of co_consts argument, for example.  Maybe I should do a benchmark and see how much the tp_traverse method would cost us.  It would be more correct to have it but I'm okay with how you've done it inline in gcmodule.c.